dali/pipeline/executor/executor.cc

```diff
@@ -356,10 +356,16 @@ int Executor<WorkspacePolicy, QueuePolicy>::InferBatchSize(
     DALI_ENFORCE(bsps[0]->NextBatchSize() == bsps[i]->NextBatchSize(),
                  "Batch size must be uniform across an iteration");
   }
-  auto batch_size = bsps[0]->NextBatchSize();
-  assert(batch_size > 0);
-  for (auto &bsp : bsps) {
-    bsp->Advance();
+  try {
+    auto batch_size = bsps[0]->NextBatchSize();
+    assert(batch_size > 0);
```

Comment:
I wouldn't enforce that - and definitely not with an assert (only debug). Once we have full batch variability, batches of 0 will be perfectly acceptable (imagine conditional execution where the predicate matches for all samples).

Reply:
Generally, I believe, that when we do have full batch variability, this function would need to be rewritten from scratch. It's because this assumes, that every `BatchSizeProvider` would return the same batch size. I've put this condition in the `assert`, cause if the batch size is nonpositive, we rather have a coding error somewhere. User can't fill the cache with `batch_size=0` data...

Reply:
Are you sure you can't just provide an empty batch in external source? Are we guarded against that?